### PR TITLE
Azure: Added linux boot time telemetry

### PR DESCRIPTION
**Pull request by iamshital**
This PR calculates the VM boot time and upload to telemetry DB.
`Boot Time in Azure = VM Turn ON --> Wala service start time.`

Please check the results listed below.

### CODE REVIEW

**Comment by iamshital** [comment]
Distro | Verified?
-- | --
Ubuntu | Yes
CentOS | Yes
Redhat | Yes
OpenSuse | Yes
SLES | Yes
OracleLinux | Yes

**Comment by iamshital** [comment]
Current Execution Logs - Ubuntu 1604
----------------------------
```
06/11/2019 13:49:41 : [INFO ] Boot time calculation started...
06/11/2019 13:49:41 : [INFO ] dos2unix: converting file .\Testscripts\Linux\CollectLogFile.sh to Unix format...
06/11/2019 13:49:41 : [INFO ] Uploading .\Testscripts\Linux\CollectLogFile.sh to lisa : 13.82.18.43, port 22 using Password authentication
06/11/2019 13:49:45 : [INFO ] Upload successful after 1 attempt
06/11/2019 13:49:45 : [INFO ] dos2unix: converting file D:\GitHub\LISAv2-2\TestResults\2019-11-06-06-45-34-4759\runtest.sh to Unix format...
06/11/2019 13:49:45 : [INFO ] Uploading D:\GitHub\LISAv2-2\TestResults\2019-11-06-06-45-34-4759\runtest.sh to lisa : 13.82.18.43, port 22 using Password authentication
06/11/2019 13:49:48 : [INFO ] Upload successful after 1 attempt
06/11/2019 13:49:48 : [INFO ] .\Tools\plink.exe -t -pw ****** -P 22 lisa@13.82.18.43 "bash CollectLogFile.sh" 
06/11/2019 13:49:50 : [INFO ] bash CollectLogFile.sh executed successfully in 2.15 seconds.
06/11/2019 13:49:50 : [INFO ] Downloading LISAv2-OneVM-SS-AX83-20190611064559-role-0-*.txt from lisa : 13.82.18.43,port 22 to D:\GitHub\LISAv2-2\TestResults\2019-11-06-06-45-34-4759 using Password authentication
06/11/2019 13:49:56 : [INFO ] Download command returned exit code 0                                                                                                                             06/11/2019 13:49:56 : [INFO ] Download successful after 1 attempt
06/11/2019 13:49:56 : [INFO ] Upload file to Azure: Success: https://xxxxxxxx.blob.core.windows.net/logs/bootPerf/LISAv2-OneVM-SS-AX83-20190611064559-role-0-waagent-636958325963078030.log.txt
06/11/2019 13:49:56 : [INFO ] Upload file to Azure: Success: https://xxxxxxxxx.blob.core.windows.net/logs/bootPerf/LISAv2-OneVM-SS-AX83-20190611064559-role-0-dmesg-636958325963078030.txt
06/11/2019 13:49:56 : [INFO ] LISAv2-OneVM-SS-AX83-20190611064559-role-0 - WALA Version = 2.2.32.2
06/11/2019 13:49:56 : [INFO ] LISAv2-OneVM-SS-AX83-20190611064559-role-0 - WALA Provision Time = 10
06/11/2019 13:49:56 : [INFO ] LISAv2-OneVM-SS-AX83-20190611064559-role-0 - Kernel Boot Time = 17 seconds
06/11/2019 13:49:56 : [INFO ] LISAv2-OneVM-SS-AX83-20190611064559-role-0 - Host Version = 14393-10.0-0-0.296
06/11/2019 13:49:56 : [INFO ] INSERT INTO LinuxDeploymentAndBootData (DateTimeUTC,TestPlatform,TestLocation,TestCaseName,SubscriptionID,SubscriptionName,ResourceGroupName,NumberOfVMsInRG,RoleName,DeploymentTime,KernelBootTime,WALAProvisionTime,HostVersion,GuestDistro,KernelVersion,LISVersion,WALAVersion,RoleSize,StorageType,CallTraces,kernelLogFile,WALAlogFile) VALUES ('2019-6-11 13:49:41','Azure','Azure','','0cc2a67a-58b9-4e4f-98a8-bfa46a28e896','Linux Integration Services TEST ONLY','LISAv2-OneVM-SS-AX83-20190611064559','1','LISAv2-OneVM-SS-AX83-20190611064559-role-0',191.0591529,17,10,'14393-10.0-0-0.296','Ubuntu 16.04.6 LTS (Xenial Xerus)','4.15.0-1046-azure','NA','2.2.32.2','Standard_A1','','No','https://xxxxxxxxxx.blob.core.windows.net/logs/bootPerf/LISAv2-OneVM-SS-AX83-20190611064559-role-0-dmesg-636958325963078030.txt','https://xxxxxxxxxx.blob.core.windows.net/logs/bootPerf/LISAv2-OneVM-SS-AX83-20190611064559-role-0-waagent-636958325963078030.log.txt')
06/11/2019 13:49:57 : [INFO ] Uploading boot data to database :  done!!
```

**Comment by iamshital** [comment]
Found PSAnalyser errors. Will fix it.

File: Libraries/Azure.psm1 [hunk 1]
```diff
@@ -2421,3 +2421,175 @@ function Add-AzureAccountFromSecretsFile {
         Raise-Exception ("XML Secrets file not provided")
     }
 }
+
+Function Upload-AzureBootAndDeploymentDataToDB ($DeploymentTime, $AllVMData, $CurrentTestData) {
```
Comment (ader1990) [inline Libraries/Azure.psm1:2448]: can you split this code into two methods: one that creates the boot information for each VM and one that uploads the information to db? This way we separate the concerns and it is easy to modify the boot info code and the db insert.

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: Libraries/Azure.psm1 (hunk 1 above)
Comment (iamshital) [reply Libraries/Azure.psm1:2448]: @ader1990 Surely.

Comment (ader1990) [reply Libraries/Azure.psm1:2448]: Thanks.

**Comment by juhlee-microsoft** [comment]
@iamshital are you only checking in CustomLIS in the DB?

File: Libraries/Azure.psm1 [hunk 2]
```diff
@@ -2421,3 +2421,175 @@ function Add-AzureAccountFromSecretsFile {
...
+
+            #region Waagent Version Checking.
+            $waagentFile = "$LogDir\$($vmData.RoleName)-waagent.log.txt"
+            $waagentStartLineNumber = (Select-String -Path $waagentFile -Pattern "$walaStartIdentifier")[0].LineNumber
+            $waagentStartLine = (Get-Content -Path $waagentFile)[$waagentStartLineNumber - 1]
+            $WALAVersion = ($waagentStartLine.Split(":")[$waagentStartLine.Split(":").Count - 1]).Trim()
+            Write-LogInfo "$($vmData.RoleName) - WALA Version = $WALAVersion"
+            #endregion
+
+            if ( ($WALAVersion -imatch "2.2.18") -or ($WALAVersion -imatch "2.2.14") ) {
```
Comment (chvalean) [inline Libraries/Azure.psm1]: please try not to hardcode versions checks like this buried in the code.
This makes maintaining it very hard.

**Review by chvalean — CHANGES_REQUESTED** [review]

File: Libraries/Azure.psm1 (hunk 2 above)
Comment (iamshital) [reply Libraries/Azure.psm1]: Older WALA versions have different message for saying - Provision is completed. So, it needs to be handled accordingly. Will check if I can take this code out into some XML.

Comment (chvalean) [reply Libraries/Azure.psm1]: or maybe just have a set of expected strings, without versions attached.
Thinking that we don't have to match a specific string to a version?

Comment (iamshital) [reply Libraries/Azure.psm1]: > or maybe just have a set of expected strings, without versions attached.
> Thinking that we don't have to match a specific string to a version?

Yes. This should work. Thanks.

**Comment by iamshital** [comment]
Results for DS1 VM size in westus2 region.
-------------------------------------------------------------
Distro | KernelBootTime | WALAProvisionTime | DeploymentTime
-- | -- | -- | --
CentOS Linux release 7.1.1503   (Core) | 18 | 19 | 109
CentOS Linux release 7.2.1511   (Core) | 10 | 17 | 81
CentOS Linux release 7.3.1611   (Core) | 10 | 15 | 95
CentOS Linux release 7.4.1708   (Core) | 20 | 14 | 124
CentOS Linux release 7.5.1804   (Core) | 21 | 18 | 114
CentOS Linux release 7.6.1810   (Core) | 14 | 10 | 79
CentOS release 6.10 (Final) | 27 | 27 | 114
CentOS release 6.5 (Final) | 18 | 27 | 127
CentOS release 6.8 (Final) | 9 | 27 | 102
CentOS release 6.9 (Final) | 33 | 25 | 128
Container Linux by CoreOS   2079.5.1 | 36 | 10 | 160
Container Linux by CoreOS   2135.2.0 | 31 | 9 | 111
Container Linux by CoreOS   2163.1.0 | 36 | 10 | 191
Oracle Linux Server 6.10 | 18 | 24 | 128
Oracle Linux Server 6.8 | 13 | 23 | 113
Oracle Linux Server 6.9 | 12 | 23 | 128
Oracle Linux Server 7.3 | 18 | 15 | 100
Oracle Linux Server 7.4 | 15 | 14 | 82
Oracle Linux Server 7.5 | 18 | 16 | 113
Oracle Linux Server 7.6 | 15 | 14 | 96
Red Hat Enterprise Linux   Server release 6.10 (Santiago) | 29 | 28 | 126
Red Hat Enterprise Linux   Server release 6.7 (Santiago) | 17 | 27 | 114
Red Hat Enterprise Linux   Server release 6.8 (Santiago) | 24 | 30 | 126
Red Hat Enterprise Linux   Server release 6.9 (Santiago) | 34 | 25 | 172
Red Hat Enterprise Linux   Server release 7.2 (Maipo) | 20 | 16 | 94
Red Hat Enterprise Linux   Server release 7.3 (Maipo) | 24 | 15 | 95
Red Hat Enterprise Linux   Server release 7.4 (Maipo) | 26 | 11 | 114
Red Hat Enterprise Linux   Server release 7.5 (Maipo) | 21 | 13 | 111
Red Hat Enterprise Linux   Server release 7.6 (Maipo) | 24 | 10 | 96
SUSE Linux Enterprise Server   11 SP4 | 45 | 37 | 207
Ubuntu 16.04.6 LTS (Xenial   Xerus) | 12 | 10 | 60
Ubuntu 18.04.2 LTS (Bionic   Beaver) | 31 | 1 | 94
Ubuntu 18.10 (Cosmic   Cuttlefish) | 30 | 1 | 96
Ubuntu 19.04 (Disco Dingo) | 38 | 2 | 111
Ubuntu 19.10 (Eoan Ermine) | 34 | 1 | 112

**I'm verifying with the team, logs and reported numbers.**

File: Libraries/Azure.psm1 (hunk 1 above)
Comment (iamshital) [reply Libraries/Azure.psm1:2448]: @ader1990 Done.

File: Libraries/Azure.psm1 (hunk 2 above)
Comment (iamshital) [reply Libraries/Azure.psm1]: @chvalean Moved the text identifiers to XML file which can be extended in future.

**Comment by iamshital** [comment]
> @iamshital are you only checking in CustomLIS in the DB?

Can you clarify your question?

I'm also adding kernel version data.

**Comment by iamshital** [comment]
By mistake clicked on close button.

File: Libraries/Azure.psm1 (hunk 1 above)
Comment (ader1990) [reply Libraries/Azure.psm1:2448]: You are still creating the sql query here, one method should not be related whatsoever with dealing sql related things.

**Review by ader1990 — APPROVED** [review]
@iamshital you can proceed with the merge to not keep the code unmerged, but we have to make sure we refactor the code in the future to have proper split code according to what it does. database management should not be combined with VM boot time calculation.

File: TestProviders/TestProvider.psm1 [hunk 3]
```diff
@@ -32,7 +32,7 @@ Class TestProvider
 	[string] $CustomLIS
 	[bool]   $ReuseVmOnFailure = $false
 
-	[object] DeployVMs([xml] $GlobalConfig, [object] $SetupTypeData, [object] $TestCaseData, [string] $TestLocation, [string] $RGIdentifier, [bool] $UseExistingRG, [string] $ResourceCleanup) {
+	[object] DeployVMs([xml] $GlobalConfig, [object] $SetupTypeData, [object] $TestCaseData, [string] $TestLocation, [string] $RGIdentifier, [bool] $UseExistingRG, [string] $ResourceCleanup, [switch] $EnableTelemetry) {
```
Comment (lizzha) [inline TestProviders/TestProvider.psm1]: Will the interface change affect other internal providers?
Can you make the EnableTelemetry a property of TestProvider instead?

**Review by lizzha — CHANGES_REQUESTED** [review]

**Comment by juhlee-microsoft** [comment]
@iamshital there are 2 open comments and waiting for your response. Please follow up in 24hr.

**Comment by iamshital** [comment]
> @iamshital there are 2 open comments and waiting for your response. Please follow up in 24hr.

@juhlee-microsoft  Addressed them.

**Comment by iamshital** [comment]
@lizzha  PR is re-verified against all distros with change in provider.

**Review by lizzha — APPROVED** [review]

**Pull request merged**